### DISCUSSION TITLE
how to find all references to an entity?

### DISCUSSION THREAD

**Discussion by ghost** [stride3d/stride#1360] [Q&A]
i have some objects that i load from a prefab, then i remove them from the scene using a script attached to them. however, they remain in memory, despite me not manually referencing them anywhere else in the game. how do i find what's referencing the entities so that i can remove them from memory? i've been trying to do this for about a week and it's really frustrating. all i know is "there's a reference to them somewhere" which is very unhelpful.

**Comment by Eideren** [comment]
With a memory profiler.
How do you know that those objects are still in memory ?
Are you certain that the GC actually did a collect since your removed those objects from the scene ?

**Reply by ghost** [reply]
I manually called GC collect once the objects were gone, for testing purposes. Then I took a snapshot with the memory profiler, there were still 300 object references in memory. The exact amount I loaded in.

**Reply by ghost** [reply]
this is after calling garbage collection and removing the entities from the scene. i'm not sure how to interpret it though.
![snapshot mem](https://user-images.githubusercontent.com/89107810/158946542-a12cbbd9-ae58-4e00-8490-2986c688fbb8.png)

**Reply by ghost** [reply]
i think i'm removing the entities while it's calculating the physics for them, which is keeping them in the physics queue, preventing GC on them. would removing their physics component before removing them fix that issue?

**Reply by manio143** [reply]
Removing an Entity from the scene results in physics components being detached and their native resources should be freed.
First of all, I don't see why do you care if those entities are still in memory, unless you can clearly see that a memory footprint increases significantly over time which would suggest a memory leak somewhere - memory managed in .NET is complicated. The screenshot shows an `EntityComponentCollection` which is something an `Entity` owns. If you're certain that GC happened in all generations and that you have removed all references to the entities from your code and after a number of frames those entities are still present in memory, then this could be a bug - but it's unlikely as we haven't had reports of memory leaks wrt creating a lot of entities (and we're talking thousands not 300).

**Reply by ghost** [reply]
if i load in 1000s of entities and then discard them, they aren't cleared from memory, memory usage does increase over time. there's never more than 1000 entities loaded at once, but the ones that were there and have since been removed are still in memory. i just set it to 300 for testing as even then they stay in memory, but doesn't increase as new entities aren't being added. i know that GC occurred after removing them as i called GC.collect() after removing the entities.

**Reply by tebjan** [reply]
this seems theoretical, the garbage collector is extremely intelligent nowadays, especially in .NET 6. it will collect when it thinks the time is right. there are also several levels of collection. so just don't bother unless you really have a real-world problem and see memory not getting removed over a long period of an hour or so when you made a release build with all optimizations enabled.

the only thing you need to know is whether any object still has a reference to these objects, if not, they will be collected.

the point of .NET is that you don't care about those kinds of issues unless they are actually affecting your software.

**Reply by ghost** [reply]
![memory snap](https://user-images.githubusercontent.com/89107810/159030638-1c276a5a-5349-47f7-9d78-2d22bfd38df9.png)
this is a snapshot of memory, taken after removing the objects and calling GC on all generations, the physics sim references a scene instance with 299 entity component collections. that would be why they're still in memory.

**Reply by manio143** [reply]
Are you sure you removed all those entities from the scene? Can you provide a minimal repro that shows the memory leak?

**Reply by ghost** [reply]
i called Entity.Scene.Entities.Remove(Entity) on all of them. they were removed from the scene. i am unsure what you mean by a minimal repro?

**Reply by manio143** [reply]
As in a smallest program that can showcase the issue. I'll try running a test on that myself later today.

**Reply by ghost** [reply]
i can try uploading it to github, then sending a link

**Reply by ghost** [reply]
https://github.com/JustAnotherGameDevWasTaken/mem-leak i put it in a zip, the whole project is only 65 MB though. i'm new to C# and stride, so i'm just trying to figure out the basics by myself as tutorials don't really sink in for me. so i've probably made a simple newbie mistake or something honestly.

**Reply by manio143** [reply]
I've modified your spawn script a little to create entities over time semi randomly
```csharp
public class Spawn : SyncScript
{
    private Prefab boxPrefab;
    public int total;

    public override void Start()
    {
        boxPrefab = Content.Load<Prefab>("Box");
        this.Setup(100);
    }

    public override void Update()
    {
        DebugText.Print($"Total: {total}, ms: {Game.UpdateTime.Total.Milliseconds}", new Int2(10, 10));
        if (Game.UpdateTime.Total.Milliseconds < 10)
            Setup(10);
    }

    private void Setup(int num)
    {
        for (int v = 0; v < num; v++)
        {
            var box = boxPrefab.Instantiate();
            foreach (var entity in box)
            {
                entity.Transform.Position.Y += 5;
                Entity.Scene.Entities.Add(entity);
                total++;
            }
        }
    }
}
```
And I don't see any memory leaking. Look at the attached video where memory used stays around 250-260MB over time  having created and removed a few thousand entities. It goes up, GC runs automatically and it goes down.

https://user-images.githubusercontent.com/10709060/159071843-f2021f95-3b9b-4952-951f-35920717902f.mp4

AFAICT you did nothing wrong and I don't understand your problem. Unless you see visible issues with memory usage there's not much point in micro profiling.

**Reply by ghost** [reply]
in the project you downloaded, i removed the materials from the cubes as i didn't think that would affect anything. turns out, the materials were the cause. if you create a new material and apply it to the cubes in the box prefab, you get the issue where memory just keeps going up.

**Reply by ghost** [reply]
i didn't check if removing the materials still had the issue, i just removed the materials thinking "this is going to be an issue with the code, it's a sample project, it doesn't need to look nice, so i'll remove the materials." i was an idiot for not checking if the materials were the cause. they are. any idea why that might be the case?

**Reply by Eideren** [reply]
Might be fixed through PR #1298 , don't have time to check

**Reply by ghost** [reply]
Found a workaround. load the entity with no material, then add it at runtime, then remove it before unloading the entity.
